_posts/2020-10-09-from-sprockets-to-webpacker.markdown

```diff
@@ -0,0 +1,315 @@
+@import "material-design-lite/src/resets/resets";
+@import "material-design-lite/src/typography/typography";
+@import "material-design-lite/src/textfield/textfield";
+@import "material-design-lite/src/slider/slider";
+
+@import "bootstrap-sass/assets/stylesheets/bootstrap";
+
+@import "vendor/assets/stylesheets/fastruby/styleguide-core.scss";
+```
+
+This is really similar, we just replaced the new packages' names and locations for each file we need.
+
+> For more details of the Styleguide changes, you can check [this Pull Request](https://github.com/fastruby/styleguide/pull/23)
+
+## Installing Webpacker for FastRuby.io
+
+The first step is easy, we have to follow the [official guide](https://github.com/rails/webpacker#installation). We don't need any special config since we are not using JS frameworks like React or Node, so the default install and config works.
+
+1 - Added the gem to the Gemfile `gem 'webpacker', '~> 5.x'`
+2 - Run `bundle` to install webpacker and its dependencies
+3 - Run `bundle exec rails webpacker:install` to initialize the configuration files
+
+After that, we'll see a new folder `app/javascript` where we'll move all our JavaScript files and a file `app/javascript/packs/application.js` which is the entry point for our code (it's similar to the `app/assets/javascript/application.js` file).
+
+There are a few more new files that we'll use to add more configuration later.
+
+## Linking the JavaScript
+
+`sprocket-rails` gem provides the `javascript_include_tag` helper to add the `<script>` tag in the head. `Webpacker` gem provides a similar helper: `javascript_pack_tag`.
+
+We'll use both files simultaneously while migrating. At the end we will remove the previous one, that way we test incremental changes.
+
+```erb
+= javascript_include_tag 'application' 
+= javascript_pack_tag 'application'
+```
+
+## Moving Files
+
+The original `application.js` includes our styleguide, some third party code and some custom js files. This is an extract:
+
+```js
+...
+//= require fastruby/styleguide
+//= require rails-ujs
+//= require contact
+//= require form
+...
+```
+
+### Simple Files
+
+We started by moving custom local files, since it requires less work. Most of the files can be moved from `app/assets/javascript` to `app/javascript/src` and import them as modules with no big changes.
+
+```js
+// app/javascript/packs/application.js
+
+import "../src/contact"
+import "../src/form"
+```
+
+The `form.js` file includes a function that should be global, so we have to fix that (webpack runs code isolated, so functions don't populate the global space by default).
+
+```js
+// app/javascript/src/form.js
+// before
+function renderValue(value) {
+...
+
+// after
+global.renderValue = function(value) {
+...
+```
+
+### 3rd Party Code
+
+We use rails-ujs to handle the remote form submission and other Rails event. Instead of moving files, for this case we have to add the `@rails/ujs` package and initialize it inside our `application.js`.
```

Comment:
```suggestion
We use rails-ujs to handle the remote form submission and other Rails events. Instead of moving files, for this case we have to add the `@rails/ujs` package and initialize it inside our `application.js`.
```